jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/EtagUtil.java

```diff
@@ -0,0 +1,251 @@
+
+    /**
+     * Create a new {@link HttpField} {@link HttpHeader#ETAG} field suitable to represent the provided Resource.
+     *
+     * @param resource the resource to represent with an Etag.
+     * @return the field if possible to create etag, or null if not possible.
+     */
+    public static HttpField createWeakEtagField(Resource resource)
+    {
+        return createWeakEtagField(resource, null);
+    }
+
+    /**
+     * Create a new {@link HttpField} {@link HttpHeader#ETAG} field suitable to represent the provided Resource.
+     *
+     * @param resource the resource to represent with an Etag.
+     * @param etagSuffix the suffix for the etag
+     * @return the field if possible to create etag, or null if not possible.
+     */
+    public static HttpField createWeakEtagField(Resource resource, String etagSuffix)
+    {
+        Path path = resource.getPath();
+        if (path == null)
+            return null;
+
+        String etagValue = EtagUtil.calcWeakEtag(path, etagSuffix);
+        if (etagValue != null)
+            return new PreEncodedHttpField(HttpHeader.ETAG, etagValue);
+        return null;
+    }
+
+    /**
+     * Calculate the weak etag for the provided {@link Path} reference.
+     *
+     * <p>
+     * This supports relative path references as well.
+     * Which can be useful to establish a reliable etag if the base changes.
+     * </p>
+     *
+     * @param path the path to calculate from
+     * @return the weak etag
+     */
+    public static String calcWeakEtag(Path path)
+    {
+        return calcWeakEtag(path, null);
+    }
+
+    /**
+     * Calculate the weak etag for the provided {@link Path} reference.
+     *
+     * @param path the path to calculate from
+     * @param suffix the suffix for the ETag
+     * @return the weak etag
+     */
+    public static String calcWeakEtag(Path path, String suffix)
+    {
+        if (path == null)
+            return null;
+
+        StringBuilder b = new StringBuilder(32);
+        b.append("W/\"");
+
+        String name = path.toString();
+        int length = name.length();
+        long lhash = 0;
+        for (int i = 0; i < length; i++)
+        {
+            lhash = 31 * lhash + name.charAt(i);
+        }
+
+        Base64.Encoder encoder = Base64.getEncoder().withoutPadding();
+        long lastModifiedMillis = -1;
+        long size = -1;
+        try
+        {
+            lastModifiedMillis = Files.getLastModifiedTime(path).toMillis();
+            size = Files.size(path);
+        }
+        catch (IOException ignored)
+        {
+            // TODO: should we ignore the inability to read last modified or size?
+        }
+        b.append(encoder.encodeToString(longToBytes(lastModifiedMillis ^ lhash)));
+        b.append(encoder.encodeToString(longToBytes(size ^ lhash)));
+        if (suffix != null)
+            b.append(suffix);
+        b.append('"');
+        return b.toString();
```

Comment:
Is there a reference for this algorithm? Is it defined by some spec? If so, please add a reference to it.

Reply:
The Etag is just a string.
The bits and pieces we use are just our take on how to create that string.
I'll add the link to the RFC9110 section in the top level javadoc.